### PR TITLE
🎉 Source Recharge: improve 'backoff' for HTTP requests

### LINKED ISSUE #7469: Source Recharge: API return mal-formated JSON need to implement retry
## Enviroment
- **Airbyte version**: 0.30.23-alpha
- **OS Version / Instance**: GCP
- **Deployment**: Docker
- **Source Connector and version**: Recharge 0.1.3
- **Destination Connector and version**: BigQuery 0.5.0
- **Severity**: Very Low / Low / Medium / **High** / Critical
- **Step where error happened**: Deploy / **Sync job** / Setup new connection / Update connector / Upgrade Airbyte

## Current Behavior
I have 10+ connections using Recharge connector. For the account with fewer data points (10k) the sync runs without any problem, but for two large accounts (500k+) the connection keep failing
image

## Expected Behavior
I expected the connector works for any size of account.

## Logs

[logs-956-0.txt](https://github.com/airbytehq/airbyte/files/7437919/logs-956-0.txt)

From logs is not clear why is failing. The error is about a mal-formated JSON reponse from the API. Maybe is a Error or a return we dont expect from their side.
```
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 - Traceback (most recent call last):
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/airbyte/integration_code/main.py", line 13, in <module>
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     launch(source, sys.argv[1:])
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 108, in launch
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     for message in source_entrypoint.run(parsed_args):
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 99, in run
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     for message in generator:
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 110, in read
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     raise e
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 106, in read
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     internal_config=internal_config,
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 136, in _read_stream
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     for record in record_iterator:
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 181, in _read_incremental
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -     for record_counter, record_data in enumerate(records, start=1):
2021-11-02 21:31:16 ERROR () LineGobbler(voidCall):65 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/streams/http/http.py", line 353, in read_records

```

## Steps to Reproduce
1.
2.
3.

## Are you willing to submit a PR?

Remove this with your answer.

### PR DESCRIPTION
**Pull request by grubberr**
Signed-off-by: Sergey Chvalyuk <grubberr@gmail.com>

## What
This Airbyte connector can fail time to time because of incorrect HTTP responses from API server.
For example you can catch such HTTP response
```
Content-Length = 337948
len(response.content) => 11922
```
As you can see 'Content-Length' of response much bigger then real response body length.
I propose to re-request such broken responses, extending current backoff mechanism.

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

### CODE REVIEW

**Review by annalvova05 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-recharge/source_recharge/api.py [hunk 1]
```diff
@@ -56,6 +56,15 @@ def get_stream_data(self, response_data: Any) -> List[dict]:
...
             return [response_data]
 
+    def should_retry(self, response: requests.Response) -> bool:
+        res = super().should_retry(response)
+        if res:
+            return res
+
+        # For some reason for large responses API server can return non-complete result
+        content_length = int(response.headers.get('Content-Length', 0))
+        return bool(content_length) and content_length > len(response.content)
```
Comment (antixar) [inline airbyte-integrations/connectors/source-recharge/source_recharge/api.py]: Can we check a code value of responses? Maybe the resource returns 204 code for this case? And is there some information or docs about empty responses?

[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-recharge/source_recharge/api.py]: yes, good point, fixed

Comment (grubberr) [reply airbyte-integrations/connectors/source-recharge/source_recharge/api.py]: No, this not related to 204 at all.
Response status = 200 but data is incomplete
Tt looks like some HTTP server layer problem
I have re-implemented and added check for status = 200 to be more explicit for this specific case

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-recharge/source_recharge/api.py]: Confirm @grubberr the response status is 200

**Review by antixar — APPROVED** [review]
looks good!

**Review by annalvova05 — APPROVED** [review]

**Comment by antixar** [comment]
@grubberr 
Please publish and merge this PR. This solution is obvious and should help to fix the bug

**Pull request merged**